lib: pixel: towards gstreamer/ffmpeg/opencv

josuah opened the pull request:
This is a proposed implementation for:

- #86669

Downstream:

- https://github.com/zephyrproject-rtos/zephyr/pull/87007
- https://github.com/zephyrproject-rtos/zephyr/pull/87008
- https://github.com/zephyrproject-rtos/zephyr/pull/88839
- https://github.com/zephyrproject-rtos/zephyr/pull/88840

Example usage:

- [samples/drivers/video/sw_pipeline/src/pipeline.c](https://github.com/zephyrproject-rtos/zephyr/blob/ff6694a667a6ae80ecc66b084a17b29c8b12d172/samples/drivers/video/sw_pipeline/src/pipeline.c)

Implementation notes:

- Functions named `pixel_<a>_to_<b>()` or  `pixel_<operation>_<a>()`, where `<a>` and `<b>` are pixel formats name with `frame`/`line`/`hist`/`avg` suffix (no reliance on `<zephyr/drivers/video.h>` or `<zephyr/drivers/display.h>` formats)
- Heavy use of `static inline`: to allow the compiler to optimize the constants away and build operations that work on whole-lines without function call inside.
- No unit tests yet, as maybe using `VIDEO_SW_GENERATOR` to build the test data was interesting?

josuah committed:
lib: pixel: Introduce a pixel and image processing library

The "pixel" library aims facilitating the implementation of all image
processing tasks, such as pixel conversion, with a focus on low-resource
environments. The processing functions scale down to per-pixel "kernels"
to line-based conversion to full streams of several frames.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
samples: lib: pixel: show usage of the pixel library

The newly introduced lib/pixel features utilities that help composing
video pipelines together for the purpose of stream processing, as well
as debug utilities.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah commented:
To test the samples on the native simulator:
```
west build -p -b native_sim samples/lib/pixel/print/ && west flash
west build -p -b native_sim samples/lib/pixel/resize/ && west flash
west build -p -b native_sim samples/lib/pixel/stats/ && west flash
west build -p -b native_sim samples/lib/pixel/stream/ && west flash
```

IAmAnkur commented on lib/pixel/resize.c:
```diff
@@ -0,0 +1,60 @@
+		size_t src_i = src_w * bits_per_pixel / BITS_PER_BYTE;
+		size_t dst_i = dst_w * bits_per_pixel / BITS_PER_BYTE;
+
+		memcpy(&dst_buf[dst_i], &src_buf[src_i], bits_per_pixel / BITS_PER_BYTE);
```
More of a suggestion:
If it is guaranteed that src and dst addresses never overlap, then we are good but for cases where src and dst addresses might overlap, i have seen memcpy leading to pixel data corruption So it is good to either handle such scenario or use safe version of memcpy (memcpy_s) which inherently checks for address overlap.

josuah replied:
This function could be called to shrink a buffer in-place, and this would lead to a bug!

josuah replied:
`memcpy_s()` seemed like unsupported, so I switched to `memmove()` to allow incrementally improve performance as use-cases shows-up. This would only copy 1~3 bytes in typical usage anyway I think.

IAmAnkur commented on drivers/video/video_sw_isp.c:
```diff
@@ -0,0 +1,460 @@
+}
+
+#ifdef CONFIG_POLL
+static int video_sw_isp_set_signal(const struct device *dev, enum video_endpoint_id ep,
```
Just a suggestion:
Seems like one of the input parameter (video_endpoint_id ep) is not getting used in function, if we do not have plan to use it, it is good to remove it.

josuah replied:
There was missing input validation on it!

ngphibang replied:
Just a side note. I think all things related to `video_endpoint_id` should be removed from the video framework. Video port / enpoint is already described in the devicetree and up to now, all the `video_endpoint_id` parameters passed in video functions are not used at all, just served as a check (which is not necessary). I plan to make a PR to remove them.

josuah replied:
A no-op everywhere in current drivers that is true... This PR uses multiple endpoints, should I try to make it single-endpoint (i.e. 1 input device, 1 output device)? Did you already have some plan you wanted to try for this scenario?

ngphibang replied:
Yes, I have.

> should I try to make it single-endpoint (i.e. 1 input device, 1 output device)?

It depends on the complexity of the real hw device that we should expose it as a single or multiple devices or not. But I think it's not really related to the discussed point here.

In the v4z framework, the `video_endpoint_id` parameter put in each API function has nothing to do with its meaning and that's why it has no usage. There are current drivers such as **mipicsi2rx** which also has multiple endpoints : ep_in is the camera sensor and ep_out is the csi (dma) but why it does not need to distinguish the ep direction like the ISP (as in the current PR) ?

So, the difference is not because a device has multiple endpoints (direction) or not, but it's because the ISP (and other devices like the PxP) falls into another catergory which (Linux) calls **m2m** device. For this kind of devices, the driver needs two separate buffer queues and hence needs to distinguish between input (which Linux falsely call "output") and output (which Linux calls "capture") side. To support this kind of devices, we can do as follow depending on each API:

- get/set_format() : add a "type" (input or output) field in struct `video_format`
- enqueue/dequeue() : add a "type" field in struct `video_buffer`
- set_stream() : add a type parameter in the function API. I see that you didn't implement this callback. **But streaming need to be done separately for each side**.
- set/get_ctrl() : nothing to do as control IDs are already different for input and output (and also, we rarely has to set controls on both input and output)
- set/get/enum_frmival() : no need.

I will push the PR this weekend. I plan to do more reworks in the framework to support m2m device as we also need it for rewriting PxP driver as said in my last comment [here](https://github.com/zephyrproject-rtos/zephyr/pull/82158#issuecomment-2625627127).

josuah replied:
Thank you for the insight, the `VIDEO_SW_ISP` proposed really has this mechanism of taking data from memory to memory.


https://github.com/zephyrproject-rtos/zephyr/issues/82939

josuah replied:
Thank you for the insight, I understand the underlying challenges a little bit better.

I do my best to make these drivers follow the APIs you introduce. :) Good thing that the PRs got split!

P.S.: `m2m` reminds of this other PR (late 2024):
- https://github.com/zephyrproject-rtos/zephyr/issues/82939

josuah commented:
force-push:
- Switch resize function from memcpy() to memmove() to allow in-place resizing (thank you @IAmAnkur)
- Add missing input validation from `video_sw_isp_set_signal` (thank you @IAmAnkur)
- Try to address CI issues at the same time, after reproducing them locally

josuah commented:
force-push:
- print the performance statistics on the stream sample
- Try to address CI issues at the same time, after reproducing them locally (different issues from last time)

josuah commented:
force-push:
- Rebase on latest main (to go past Zephyr 4.1 and base other PRs on this)

josuah commented:
force-push:
- Another attempt to fix CI
- 
I do not mean to spam, but this was working well with `west twister -p m2gl025_miv/miv -s sample.pixel.stats --no-detailed-test-id` locally, unlike CI. Hope this is the last!

ngphibang commented:
I think there are 3 distinct things in this PR

- Introduction of "libpixel"
- Introduction of "video_sw_generator_ISP"
- Adding a video API to collect statistic.

Should we make 3 separate PRs for each of them (without copying the commits from one PR to another) to facilitate the review and merge ? I know that a PR can depend on others but we can explicitly describe the dependency in the description (at least libpixel is indepedent from others, I guess). This way, we will not be confused when reviewing common commits. I know that there are not an optimal / unique strategy of making dependent PRs in github but it seems better doing this way ?

josuah commented:
@ngphibang
Good point! Here is the new map of RFCs and associated PRs:

Data processing use-case:

- https://github.com/zephyrproject-rtos/zephyr/issues/86669
- https://github.com/zephyrproject-rtos/zephyr/pull/86671
- https://github.com/zephyrproject-rtos/zephyr/pull/87007

Image statistics use-case:

- https://github.com/zephyrproject-rtos/zephyr/issues/85457
- https://github.com/zephyrproject-rtos/zephyr/pull/85811
- https://github.com/zephyrproject-rtos/zephyr/pull/87008

josuah committed:
tests: lib: pixel: add unit tests for format conversion

The tests use data generated by the ffmpeg command line utilty in order
to avoid bias by having the same person implementing the tests and the
source code.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah commented:
force-push:

- More consistent naming scheme for macros (`STREAM_XXX_TO_YYY` -> `XXXSTREAM_TO_YYYSTREAM` like function names)
- Improve the doc grouping of functions (this was only documenting a single line)
- Do not expose the single-pixel conversion functions, which:
  - Greatly reduce the number of functions
  - Still permits to convert a single pixel by specifying a width of 1 to the line conversion functions
  - Simplifies handling of formats that are not defined at a single-pixel level (i.e. YUYV)
- Introduce the RGB332 format.
- Rebase on latest `main`

josuah commented:
force-push:
- Enable UTF-8 back again
- Introduce a choice Kconfig to select the printing method (printf, printk, shell_print), including "none" to skip printing (CI timed-out due to large prints)
- Switch docstrings from @defgroup to @copydetails to document functions in batch, it seemed more intuitive and appropriate.

josuah committed:
MAINTAINERS: Add a "Pixel library" area

Add "Pixel library" area, which covers the newly introduced lib/pixel
and associated tests and samples.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah commented:
force-push:
- Syntax sugar: pipeline [one-liners](https://github.com/zephyrproject-rtos/zephyr/pull/86671/files#diff-26c9cb29ca88d530039dbbadd57824354209f59ce9a093ef4524efa136847900R51-R56) in C (+ macros)
- Add `__weak` attribute to line conversion functions to allow optimized version to be opted-in (i.e. SIMD).
- minor bugfixes

josuah commented:
Force-push:

- CI bugfixes
- Rebase on main
- Add kernel processing for both convolutional kernels (i.e. sharpen mask) or custom kernels (i.e. "median" for denoising).

Unit test result for gaussian Blur kernel:
![scrot_20250329_193647_222x443](https://github.com/user-attachments/assets/e589878e-51d6-4bdc-876a-da75f0f3469b)

Unit test result for median (denoise) kernel:
![scrot_20250329_192121_222x442](https://github.com/user-attachments/assets/c446a266-6cf2-4832-bf1c-d00ab313fea7)

josuah commented:
force-push:

- I could not reproduce job [39768403923](https://github.com/zephyrproject-rtos/zephyr/actions/runs/14175714305/job/39768403923?pr=86671) locally: it runs well when I test on this sample platform, so removing it for now. All tests are kept and pass CI.

New features will be proposed on other branches to reduce review burden, now using 1240bfdcd7707b5c971adf57a6e3e9753b16908d as stable checkpoint.

josuah commented:
Force-push:

- Rebase on top of latest `main` to fix a conflict.

josuah commented:
force-push:

- Minor bugfixes and API fine-tuning

josuah commented:
Force-push:

- Fix typo in copyright statement.
- Add YUV24 <-> RGB24 functions implemented but missing from header.
- Add YUV24 <-> YUYV not implemented (only direct RGB24 <-> YUYV was there).
- Add tests for all YUV24 functions above.

josuah committed:
drivers: video: add the YUV24 format

Introduce the YUV24 format following the Linux V4L2 naming and
fourcc definition.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
drivers: video: add the RGB332 format

Introduce the RGB332 format following the Linux V4L2 naming and
fourcc definition.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
drivers: video: fix the RGB565X format's byte per pixel

THe RGB565X (big-endian) did not have an entry for the function
video_byte_per_pixel(). Fix it by adding the missing entry.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah commented:
It seems that we have a bit of time before reviews, so I went ahead and implemented the python-style API.

After all, if we *can* have an API that's easy like a script but is light even on microcontrollers and makes use of all hardware accelerators available, why shouldn't we have it... :) 

```c
struct pixel_image img;

/* Open that buffer as an image type */
pixel_image_from_buffer(&img, rgb24frame_in, sizeof(rgb24frame_in), 32, 8, VIDEO_PIX_FMT_RGB24);

/* Turn it into a tall vertical image, now displeasant "banding" artifacts appear */
pixel_image_resize(&img, 5, 40);

/* Try to attenuate it with a blur effect (comment this line to see the difference) */
pixel_image_kernel(&img, PIXEL_KERNEL_GAUSSIAN_BLUR, 5);

/* Stretch the gradient horizontally over the entire width of the output buffer */
pixel_image_resize(&img, 120, 40);

/* Save the image into the output buffer and check for errors */
pixel_image_to_buffer(&img, rgb24frame_out, sizeof(rgb24frame_out));
if (img.err != 0) {
        return img.err;
}

/* Now that the image is exported, we can print it */
pixel_image_print_truecolor(&img);
```

It is only about time that [ARM Helium SIMD instructions](https://community.arm.com/arm-community-blogs/b/tools-software-ides-blog/posts/start-early-development-on-arm-cortex-m55-processor) and [custom 2.5D GPUs](https://github.com/zephyrproject-rtos/zephyr/pull/88840)... are integrated

josuah commented:
force-push:
- doc typo fix
- while at it, add missing Y8/GREY format support (useful for downstream PRs)
- add convenience functions to convert data from/to video buffers

josuah commented:
Force-push:
- Add missing Y8/GREY support (used by downstream PR)
- Doc typo fixes
- Add wrappers to convert video buffers

loicpoulain commented on include/zephyr/pixel/image.h:
```diff
@@ -0,0 +1,203 @@
+int pixel_image_add_uncompressed(struct pixel_image *img, const struct pixel_operation *template);
+
+/**
+ * @brief Perform all the processing added to the
```
unfinished sentence, is it triggering the processing pipeline?

josuah replied:
Good catch! I will check if I have more up to date documentation somewhere else and patch it.

loicpoulain commented on include/zephyr/pixel/image.h:
```diff
@@ -0,0 +1,203 @@
+ * @note This is a low-level function only needed to implement new operations.
+ *
+ * The operation step will not be processed immediately, but rather added to a linked list of
+ * operations that will be performed at runtime.
```
It will be added as latest operation, then maybe it should be clearer, naming it `pixel_image_append_operation()`?

josuah replied:
Absolutely! :pray:

loicpoulain commented on lib/pixel/bayer.c:
```diff
@@ -0,0 +1,341 @@
+	}
+}
+
+static inline void pixel_op_bayer_to_rgb24_2x2(struct pixel_operation *op, fn_2x2_t *fn0,
```
why explicit inline of that function?

josuah replied:
This is under-documented, but here is the idea.

[This function](https://github.com/zephyrproject-rtos/zephyr/pull/86671/files/a1f7efc8203862c03241a5fc6966ec38889a62c1..af9cb063a124d764a73e2fdcc3d39bfb4024d4f1#diff-6d35bc855163fe1d51efe4cdaa0ef2aa55f133434d959557778f596fc3ab57c1R51) is what performs the actual pixel operation of debayer

```c
static inline void pixel_rggb8_to_rgb24_3x3(const uint8_t rgr0[3], const uint8_t gbg1[3],
					    const uint8_t rgr2[3], uint8_t rgb24[3])
{
	rgb24[0] = ((uint16_t)rgr0[0] + rgr0[2] + rgr2[0] + rgr2[2]) / 4;
	rgb24[1] = ((uint16_t)rgr0[1] + gbg1[2] + gbg1[0] + rgr2[1]) / 4;
	rgb24[2] = gbg1[1];
}
```

This is called in a loop here:

```c
__weak void pixel_line_rggb8_to_rgb24_3x3(const uint8_t *i0, const uint8_t *i1,
					     const uint8_t *i2, uint8_t *o0, uint16_t w)
{
	__ASSERT_NO_MSG(w >= 4 && w % 2 == 0);
	uint8_t il[3][3] = FOLD_L_3X3(i0, i1, i2);
	uint8_t ir[3][3] = FOLD_R_3X3(i0, i1, i2, w);

	pixel_grbg8_to_rgb24_3x3(il[0], il[1], il[2], &o0[0]);
	for (size_t i = 0, o = 3; i + 4 <= w; i += 2, o += 6) {
		pixel_rggb8_to_rgb24_3x3(&i0[i + 0], &i1[i + 0], &i2[i + 0], &o0[o + 0]);
		pixel_grbg8_to_rgb24_3x3(&i0[i + 1], &i1[i + 1], &i2[i + 1], &o0[o + 3]);
	}
	pixel_rggb8_to_rgb24_3x3(ir[0], ir[1], ir[2], &o0[w * 3 - 3]);
}
```

This builds a "tight loop" where as much execution as possible should be removed for best performance. Other implementations (i.e. SIMD ones) actually implement debayer in just one big function... possibly for that reason. I wanted to split for deduplicating code and make it more readable without the cost of introducing function calls everywhere, so was encouraging the compiler to spend the extra flash storage to allow a tighter loop where constants can be combined/eliminated too.

josuah replied:
`pixel_op_bayer_to_rgb24_2x2()` is not so tight of a loop though, only called once per line... I feel like optimizing prematurely and I've not gotten time to do proper benchmarks.
